src/Signature/SignatureV4.php

```diff
@@ -324,11 +324,11 @@ private function getCanonicalizedQuery(array $query)
         ksort($query);
         foreach ($query as $k => $v) {
             if (!is_array($v)) {
-                $qs .= rawurlencode($k) . '=' . rawurlencode($v) . '&';
+                $qs .= rawurlencode($k ?? '') . '=' . rawurlencode($v ?? '') . '&';
```

Comment:
The ?? operator was added in 7.0, but the SDK supports back to 5.5, so this will probably make the tests that are running fail.  Even if it doesn't, it's a non backward-compatible change.  What circumstance is this needed in though? 
I need to look into this snippet of code more to make sure, but it looks like it would never be null since it's iterating through an array and getting the key/value pairs, where $k is the key.  When would a key in a query be null?

Reply:
the ?? operator could be fixed in a backward compatible way for sure. Unfortunately don't have live debug data for this case now but trace looks like:
```
/var/www/html/vendor/aws/aws-sdk-php/src/Signature/SignatureV4.php:327 rawurlencode(): Passing null to parameter #1 ($string) of type string is deprecated
/var/www/html/vendor/aws/aws-sdk-php/src/Signature/SignatureV4.php:327
/var/www/html/vendor/aws/aws-sdk-php/src/Signature/SignatureV4.php:285
/var/www/html/vendor/aws/aws-sdk-php/src/Signature/SignatureV4.php:103
/var/www/html/vendor/aws/aws-sdk-php/src/Signature/S3SignatureV4.php:46
/var/www/html/vendor/aws/aws-sdk-php/src/Middleware.php:126
/var/www/html/vendor/guzzlehttp/promises/src/Promise.php:204
/var/www/html/vendor/guzzlehttp/promises/src/Promise.php:153
/var/www/html/vendor/guzzlehttp/promises/src/TaskQueue.php:48
/var/www/html/vendor/guzzlehttp/promises/src/Promise.php:248
```

that's with PHP8.1 - https://www.php.net/manual/en/migration81.deprecated.php#migration81.deprecated.core.null-not-nullable-internal

Reply:
agree that $k could not be null. But $v could be
https://github.com/guzzle/psr7/blob/c55d23ab371b472342bee943a8bdb9c3cf963a18/src/Query.php#L45
most likely that "Passing null to parameter" recorded for second occurrence of `rawurlencode` in that line

Reply:
https://github.com/aws/aws-sdk-php/blob/8c4863a369fd3aab630270ea09b576de81887a3b/src/Signature/SignatureV4.php#L402

Reply:
Hi @karyna-tsymbal-atwix !
Could you please do corresponding changes ? (I don't think I can commit\push to PR branch)

Reply:
Hello @xmav  @SamRemis  , I've made changes according to the discussion above